core/src/main/java/com/google/cloud/sql/core/JndiDnsResolver.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.stream.Collectors;
+import javax.naming.NameNotFoundException;
+import javax.naming.NamingException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.InitialDirContext;
+
+/** Implements DnsResolver using the Java JNDI built-in DNS directory. */
+class JndiDnsResolver implements DnsResolver {
+  private final String jndiPrefix;
+
+  /** Creates a resolver using the system DNS settings. */
+  JndiDnsResolver() {
+    this.jndiPrefix = "dns:";
+  }
+
+  /**
+   * Creates a DNS resolver that uses a specific DNS server.
+   *
+   * @param dnsServer the DNS server hostname
+   * @param port the DNS server port (DNS servers usually use port 53)
+   */
+  JndiDnsResolver(String dnsServer, int port) {
+    this.jndiPrefix = "dns://" + dnsServer + ":" + port + "/";
+  }
+
+  /**
+   * Returns DNS records for a domain name, sorted alphabetically.
```

Comment:
For a domain name, we should only get 1 TXT record if one exists. Why do we have Collection<String> for the return type?

Reply:
We expect exactly 1 DNS record for an instance. However, it is possible that the connector could receive 0, 1, or many DNS records. Other application logic will gracefully handle cases where the number of responses is not 1 as expected. This class is only responsible for resolving DNS and reporting the results. Thus its data model is in line with the DNS protocol, returning a collection that may contain 0 or more responses.